// @backstage/plugin-scaffolder-react/dist/next/components/TemplateOutputs/DefaultTemplateOutputs.esm.js
import { InfoCard, MarkdownContent } from '@backstage/core-components';
import Box from '@material-ui/core/Box';
import Paper from '@material-ui/core/Paper';
import React, { useState, useEffect, useMemo } from 'react';
import { LinkOutputs } from './LinkOutputs.esm.js';
import { TextOutputs } from './TextOutputs.esm.js';

const DefaultTemplateOutputs = (props) => {
  var _a, _b;
  const { output } = props;
  const [textOutputIndex, setTextOutputIndex] = useState(
    void 0
  );
  useEffect(() => {
    if (textOutputIndex === void 0 && (output == null ? void 0 : output.text)) {
      const defaultIndex = output.text.findIndex(
        (t) => t.default
      );
      setTextOutputIndex(defaultIndex >= 0 ? defaultIndex : 0);
    }
  }, [textOutputIndex, output]);
  const textOutput = useMemo(
    () => {
      var _a2;
      return textOutputIndex !== void 0 ? (_a2 = output == null ? void 0 : output.text) == null ? void 0 : _a2[textOutputIndex] : null;
    },
    [output, textOutputIndex]
  );
  if (!output) {
    return null;
  }
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(Box, { paddingBottom: 2 }, /* @__PURE__ */ React.createElement(Paper, null, /* @__PURE__ */ React.createElement(Box, { padding: 2, justifyContent: "center", display: "flex", gridGap: 16 }, /* @__PURE__ */ React.createElement(
    TextOutputs,
    {
      output,
      index: textOutputIndex,
      setIndex: setTextOutputIndex
    }
  ), /* @__PURE__ */ React.createElement(LinkOutputs, { output })))), textOutput ? /* @__PURE__ */ React.createElement(Box, { paddingBottom: 2 }, /* @__PURE__ */ React.createElement(
    InfoCard,
    {
      title: (_a = textOutput.title) != null ? _a : "Text Output",
      noPadding: true,
      titleTypographyProps: { component: "h2" }
    },
    /* @__PURE__ */ React.createElement(Box, { padding: 2, height: "100%" }, /* @__PURE__ */ React.createElement(MarkdownContent, { content: (_b = textOutput.content) != null ? _b : "" }))
  )) : null);
};

export { DefaultTemplateOutputs };
