// @backstage/app-defaults/dist/createApp.esm.js
import { apis } from './defaults/apis.esm.js';
import { components } from './defaults/components.esm.js';
import { icons } from './defaults/icons.esm.js';
import { themes } from './defaults/themes.esm.js';
import { createSpecializedApp } from '@backstage/core-app-api';

function createApp(options) {
  var _a, _b, _c, _d;
  return createSpecializedApp({
    ...options,
    apis: (_a = options == null ? void 0 : options.apis) != null ? _a : [],
    bindRoutes: options == null ? void 0 : options.bindRoutes,
    components: {
      ...components,
      ...options == null ? void 0 : options.components
    },
    configLoader: options == null ? void 0 : options.configLoader,
    defaultApis: apis,
    icons: {
      ...icons,
      ...options == null ? void 0 : options.icons
    },
    plugins: (_b = options == null ? void 0 : options.plugins) != null ? _b : [],
    featureFlags: (_c = options == null ? void 0 : options.featureFlags) != null ? _c : [],
    themes: (_d = options == null ? void 0 : options.themes) != null ? _d : themes
  });
}

export { createApp };
