// @backstage/frontend-plugin-api/dist/wiring/createPlugin.esm.js
import { resolveExtensionDefinition } from './resolveExtensionDefinition.esm.js';

function createPlugin(options) {
  var _a, _b, _c, _d;
  const extensions = ((_a = options.extensions) != null ? _a : []).map(
    (def) => resolveExtensionDefinition(def, { namespace: options.id })
  );
  const extensionIds = extensions.map((e) => e.id);
  if (extensionIds.length !== new Set(extensionIds).size) {
    const duplicates = Array.from(
      new Set(
        extensionIds.filter((id, index) => extensionIds.indexOf(id) !== index)
      )
    );
    throw new Error(
      `Plugin '${options.id}' provided duplicate extensions: ${duplicates.join(
        ", "
      )}`
    );
  }
  return {
    $$type: "@backstage/BackstagePlugin",
    version: "v1",
    id: options.id,
    routes: (_b = options.routes) != null ? _b : {},
    externalRoutes: (_c = options.externalRoutes) != null ? _c : {},
    featureFlags: (_d = options.featureFlags) != null ? _d : [],
    extensions,
    toString() {
      return `Plugin{id=${options.id}}`;
    }
  };
}

export { createPlugin };
