// @backstage/frontend-plugin-api/dist/routing/ExternalRouteRef.esm.js
import { RouteRefImpl } from './RouteRef.esm.js';
import { describeParentCallSite } from './describeParentCallSite.esm.js';

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, key + "" , value);
  return value;
};
class ExternalRouteRefImpl extends RouteRefImpl {
  constructor(optional, params = [], defaultTarget, creationSite) {
    super(params, creationSite);
    this.optional = optional;
    this.params = params;
    this.defaultTarget = defaultTarget;
    __publicField(this, "$$type", "@backstage/ExternalRouteRef");
  }
  getDefaultTarget() {
    return this.defaultTarget;
  }
}
function createExternalRouteRef(options) {
  return new ExternalRouteRefImpl(
    Boolean(options == null ? void 0 : options.optional),
    options == null ? void 0 : options.params,
    options == null ? void 0 : options.defaultTarget,
    describeParentCallSite()
  );
}

export { createExternalRouteRef };
