// @backstage/plugin-scaffolder-react/dist/next/components/Stepper/ErrorListTemplate/errorListTemplate.esm.js
import React from 'react';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Paper from '@material-ui/core/Paper';
import { makeStyles, createStyles } from '@material-ui/core/styles';
import ErrorIcon from '@material-ui/icons/Error';

const useStyles = makeStyles(
  (_theme) => createStyles({
    list: {
      width: "100%"
    },
    text: {
      textWrap: "wrap"
    }
  })
);
const ErrorListTemplate = ({ errors }) => {
  const classes = useStyles();
  return /* @__PURE__ */ React.createElement(Paper, null, /* @__PURE__ */ React.createElement(List, { dense: true, className: classes.list }, errors.map((error, index) => /* @__PURE__ */ React.createElement(ListItem, { key: index }, /* @__PURE__ */ React.createElement(ListItemIcon, null, /* @__PURE__ */ React.createElement(ErrorIcon, { color: "error" })), /* @__PURE__ */ React.createElement(
    ListItemText,
    {
      classes: { primary: classes.text },
      primary: error.stack
    }
  )))));
};

export { ErrorListTemplate };
