// @backstage/plugin-scaffolder/dist/components/fields/SecretInput/SecretInput.esm.js
import React from 'react';
import { useTemplateSecrets } from '@backstage/plugin-scaffolder-react';
import { ScaffolderField } from '@backstage/plugin-scaffolder-react/alpha';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';

const SecretInput = (props) => {
  var _a;
  const { setSecrets, secrets } = useTemplateSecrets();
  const {
    name,
    onChange,
    schema: { title, description },
    rawErrors,
    disabled,
    errors,
    required
  } = props;
  return /* @__PURE__ */ React.createElement(
    ScaffolderField,
    {
      rawErrors,
      rawDescription: description,
      disabled,
      errors,
      required
    },
    /* @__PURE__ */ React.createElement(InputLabel, { htmlFor: title }, title),
    /* @__PURE__ */ React.createElement(
      Input,
      {
        id: title,
        "aria-describedby": title,
        onChange: (e) => {
          var _a2, _b;
          onChange(Array((_a2 = e.target) == null ? void 0 : _a2.value.length).fill("*").join(""));
          setSecrets({ [name]: (_b = e.target) == null ? void 0 : _b.value });
        },
        value: (_a = secrets[name]) != null ? _a : "",
        type: "password",
        autoComplete: "off"
      }
    )
  );
};

export { SecretInput };
