// @backstage/plugin-catalog/dist/components/CatalogTable/PaginatedCatalogTable.esm.js
import React from 'react';
import { Table } from '@backstage/core-components';
import { CatalogTableToolbar } from './CatalogTableToolbar.esm.js';

function PaginatedCatalogTable(props) {
  const { columns, data, next, prev, title, isLoading, options } = props;
  return /* @__PURE__ */ React.createElement(
    Table,
    {
      title: isLoading ? "" : title,
      columns,
      data,
      options: {
        ...options,
        // These settings are configured to force server side pagination
        paginationPosition: "both",
        pageSizeOptions: [],
        showFirstLastPageButtons: false,
        pageSize: Number.MAX_SAFE_INTEGER,
        emptyRowsWhenPaging: false
      },
      onPageChange: (page) => {
        if (page > 0) {
          next == null ? void 0 : next();
        } else {
          prev == null ? void 0 : prev();
        }
      },
      components: {
        Toolbar: CatalogTableToolbar
      },
      page: prev ? 1 : 0,
      totalCount: next ? Number.MAX_VALUE : Number.MAX_SAFE_INTEGER,
      localization: { pagination: { labelDisplayedRows: "" } }
    }
  );
}

export { PaginatedCatalogTable };
