// @backstage/frontend-plugin-api/dist/extensions/createSignInPageExtension.esm.js
import React, { lazy } from 'react';
import { ExtensionBoundary } from '../components/ExtensionBoundary.esm.js';
import { createExtensionDataRef } from '../wiring/createExtensionDataRef.esm.js';
import { createExtension } from '../wiring/createExtension.esm.js';

function createSignInPageExtension(options) {
  var _a;
  return createExtension({
    kind: "sign-in-page",
    namespace: options == null ? void 0 : options.namespace,
    name: options == null ? void 0 : options.name,
    attachTo: (_a = options.attachTo) != null ? _a : { id: "app/root", input: "signInPage" },
    configSchema: options.configSchema,
    inputs: options.inputs,
    disabled: options.disabled,
    output: {
      component: createSignInPageExtension.componentDataRef
    },
    factory({ config, inputs, node }) {
      const ExtensionComponent = lazy(
        () => options.loader({ config, inputs }).then((component) => ({ default: component }))
      );
      return {
        component: (props) => /* @__PURE__ */ React.createElement(ExtensionBoundary, { node, routable: true }, /* @__PURE__ */ React.createElement(ExtensionComponent, { ...props }))
      };
    }
  });
}
((createSignInPageExtension2) => {
  createSignInPageExtension2.componentDataRef = createExtensionDataRef("core.sign-in-page.component");
})(createSignInPageExtension || (createSignInPageExtension = {}));

export { createSignInPageExtension };
