// @backstage/plugin-catalog/dist/components/EntityLabelsCard/EntityLabelsEmptyState.esm.js
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import React from 'react';
import { CodeSnippet } from '@backstage/core-components';

const ENTITY_YAML = `metadata:
  name: example
  labels:
    javaVersion: 1.2.3`;
const useStyles = makeStyles(
  (theme) => ({
    code: {
      borderRadius: 6,
      margin: theme.spacing(2, 0),
      background: theme.palette.type === "dark" ? "#444" : theme.palette.common.white
    }
  }),
  { name: "PluginCatalogEntityLabelsEmptyState" }
);
function EntityLabelsEmptyState() {
  const classes = useStyles();
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(Typography, { variant: "body1" }, "No labels defined for this entity. You can add labels to your entity YAML as shown in the highlighted example below:"), /* @__PURE__ */ React.createElement("div", { className: classes.code }, /* @__PURE__ */ React.createElement(
    CodeSnippet,
    {
      text: ENTITY_YAML,
      language: "yaml",
      showLineNumbers: true,
      highlightedNumbers: [3, 4, 5, 6],
      customStyle: { background: "inherit", fontSize: "115%" }
    }
  )), /* @__PURE__ */ React.createElement(
    Button,
    {
      variant: "contained",
      color: "primary",
      target: "_blank",
      href: "https://backstage.io/docs/features/software-catalog/descriptor-format#labels-optional"
    },
    "Read more"
  ));
}

export { EntityLabelsEmptyState };
