// @backstage/core-components/dist/components/DependencyGraph/Node.esm.js
import React from 'react';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { DefaultNode } from './DefaultNode.esm.js';
import { NODE_TEST_ID } from './constants.esm.js';

const useStyles = makeStyles(
  (theme) => ({
    node: {
      transition: `${theme.transitions.duration.shortest}ms`
    }
  }),
  { name: "BackstageDependencyGraphNode" }
);
const renderDefault = (props) => /* @__PURE__ */ React.createElement(DefaultNode, { ...props });
function Node({
  render = renderDefault,
  setNode,
  node
}) {
  const { width, height, x = 0, y = 0 } = node;
  const nodeProps = node;
  const classes = useStyles();
  const nodeRef = React.useRef(null);
  React.useLayoutEffect(() => {
    if (nodeRef.current) {
      let { height: renderedHeight, width: renderedWidth } = nodeRef.current.getBBox();
      renderedHeight = Math.round(renderedHeight);
      renderedWidth = Math.round(renderedWidth);
      if (renderedHeight !== height || renderedWidth !== width) {
        setNode(node.id, {
          ...node,
          height: renderedHeight,
          width: renderedWidth
        });
      }
    }
  }, [node, width, height, setNode]);
  return /* @__PURE__ */ React.createElement(
    "g",
    {
      ref: nodeRef,
      "data-testid": NODE_TEST_ID,
      className: classes.node,
      transform: `translate(${x - width / 2},${y - height / 2})`
    },
    render({ node: nodeProps })
  );
}

export { Node };
