// @backstage/core-components/dist/layout/HeaderLabel/HeaderLabel.esm.js
import Grid from '@material-ui/core/Grid';
import { makeStyles, alpha } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import React from 'react';
import { Link } from '../../components/Link/Link.esm.js';

const useStyles = makeStyles(
  (theme) => ({
    root: {
      textAlign: "left"
    },
    label: {
      color: theme.page.fontColor,
      fontWeight: theme.typography.fontWeightBold,
      letterSpacing: 0,
      fontSize: theme.typography.fontSize,
      marginBottom: theme.spacing(1) / 2,
      lineHeight: 1
    },
    value: {
      color: alpha(theme.page.fontColor, 0.8),
      fontSize: theme.typography.fontSize,
      lineHeight: 1
    }
  }),
  { name: "BackstageHeaderLabel" }
);
const HeaderLabelContent = ({
  value,
  className,
  typographyRootComponent
}) => {
  return /* @__PURE__ */ React.createElement(
    Typography,
    {
      component: typographyRootComponent != null ? typographyRootComponent : typeof value === "string" ? "p" : "span",
      className
    },
    value
  );
};
function HeaderLabel(props) {
  const { label, value, url, contentTypograpyRootComponent } = props;
  const classes = useStyles();
  const content = /* @__PURE__ */ React.createElement(
    HeaderLabelContent,
    {
      className: classes.value,
      value: value || "<Unknown>",
      typographyRootComponent: contentTypograpyRootComponent
    }
  );
  return /* @__PURE__ */ React.createElement(Grid, { item: true }, /* @__PURE__ */ React.createElement(Typography, { component: "span", className: classes.root }, /* @__PURE__ */ React.createElement(Typography, { className: classes.label }, label), url ? /* @__PURE__ */ React.createElement(Link, { to: url }, content) : content));
}

export { HeaderLabel };
